File: leetcode/2001-2999/2116-check-if-a-parentheses-string-can-be-valid/CheckIfAParenthesesStringCanBeValid.ts
function canBeValid(s: string, locked: string): boolean {
  if (s.length % 2 === 1) {
    return false;
  }

  let left = 0;
  let right = 0;

  for (let i = 0; i < s.length; i++) {
    if (locked[i] === "0") {
      left++;
      right++;
    } else {
      if (s[i] === "(") {
        left++;
      } else {
        left--;
      }

      if (s[s.length - 1 - i] === ")") {
        right++;
      } else {
        right--;
      }

      if (left < 0 || right < 0) {
        return false;
      }
    }
  }

  return true;
};

console.log(canBeValid("))()))", "010100"))

File: leetcode/2001-3000/2101-2200/2116-check-if-a-parentheses-string-can-be-valid/CheckIfAParenthesesStringCanBeValid.java
public class CheckIfAParenthesesStringCanBeValid {
  public static void main(String[] args) {
    System.out.println(new Solution().canBeValid("))()))", "110100"));
  }
}

class Solution {
  public boolean canBeValid(String s, String locked) {
    if (s.length() % 2 == 1) {
      return false;
    }

    int totalValidCharacters = 0;

    for (int i = 0; i < s.length(); i++) {
      if (s.charAt(i) == '(' || locked.charAt(i) == '0') {
        totalValidCharacters++;
      } else if (totalValidCharacters > 0) {
        totalValidCharacters--;
      } else {
        return false;
      }
    }

    totalValidCharacters = 0;

    for (int i = s.length() - 1; i >= 0; i--) {
      if (s.charAt(i) == ')' || locked.charAt(i) == '0') {
        totalValidCharacters++;
      } else if (totalValidCharacters > 0) {
        totalValidCharacters--;
      } else {
        return false;
      }
    }

    return true;
  }
}
